### ISSUE TITLE
qrDecompose.angle does not match obj.angle with flipX applied

### ISSUE THREAD

**Issue by cmawhorter** [fabricjs/fabric.js#5079]
## Version
2.3.3

## Test Case
Go to http://fabricjs.com/matrix-transformation and run this script in console:

```js
var obj = __canvas.item(0);
function testObject() {
  var transformValues = fabric.util.qrDecompose(obj.calcTransformMatrix());
  var match = obj.angle === transformValues.angle;
  console.log(match ? 'yes' : 'NO', obj.angle, transformValues);
}
testObject(); // obj.angle and matrix.angle match
obj.flipX = true;
__canvas.requestRenderAll();
setTimeout(testObject, 100); // obj.angle correctly zero. matrix.angle is now 180
```

## Steps to reproduce

Run the script above or:

1. add an object to canvas.  obj.angle will be zero.  calcTransformMatrix + qrDecompose angle will be zero
1. obj.flipX = true; requestRenderAll()
1. obj.angle remains zero.  qrDecompose now returns 180

## Expected Behavior

`fabric.util.qrDecompose(obj.calcTransformMatrix()).angle` and `obj.angle` should both equal zero, or qrDecompose shouldn't return an angle property if it can't return an accurate value.

## Actual Behavior

After the flip is applied, the result from qrDecompose is 180. 

If qrDecompose can't be used this way, what is the recommended way to get the ungrouped object properties from a selection's object:modified event?

**Comment by asturur** [comment]
qrDecompose is accurate enough.
The problem behind it is that it has to return:
6 properties from 6 numbers that are not univocally related.
Numbers are added and multiplied when forming the transform matrix and going back is not possible.
We find one of the possible solutions.

An object that has been flipped on both axis ( x and y ) is indeed rotated of 180 degree, you can't tell the difference.

qrDecompose is the best thing that happened to this library, it solved like 10 long standing bugs, and the flip+rotation is the only drawback we have to keep.

If you feel like fixing it and have some idea about it, please let's talk about it.

**Comment by cmawhorter** [comment]
> you can't tell the difference.

you can't visually tell the difference of the image, but the issue seems to impact controls and also anything that depends on that number being accurate.  

additionally, in my example it was flipX only -- and not flipY.  that's why i discovered this issue in the first place.  user objects were randomly flipping.

> If you feel like fixing it and have some idea about it, please let's talk about it.

in the documentation it says it works with matrix values so it probably shouldn't return an angle in the first place.

i'd love it if i could work with two properties: transformMatrix and angle.  working with left/top/scale and the rest has always been buggy for me.

**Comment by melvynhills** [comment]
I could fix this behaviour for my needs with these changes (not thoroughly tested yet).

As soon as flip values are changed, subtracting 180° from the angle and set back the original flip values seems to do the trick.

```javascript
fabric.Group.prototype.realizeTransform = function(object) {
  var matrix = object.calcTransformMatrix(),
      options = fabric.util.qrDecompose(matrix),
      center = new fabric.Point(options.translateX, options.translateY),
  // custom code start ------------------------------------------------------------------
      originalFlipX = object.flipX,
      originalFlipY = object.flipY;
  // custom code end ------------------------------------------------------------------

  object.flipX = false;
  object.flipY = false;
  object.set('scaleX', options.scaleX);
  object.set('scaleY', options.scaleY);
  object.skewX = options.skewX;
  object.skewY = options.skewY;

  // custom code start ------------------------------------------------------------------
  if (originalFlipX !== object.flipX || originalFlipY !== object.flipY) {
    object.flipX = originalFlipX;
    object.flipY = originalFlipY;
    options.angle -= 180;
  }
  // custom code end ------------------------------------------------------------------

  object.angle = options.angle;
  object.setPositionByOrigin(center, 'center', 'center');

  return object;
};

```

**Comment by asturur** [comment]
are we sure that either a flip or the other will always give a 180 change? if that would be only change needed, we could also implement it.

A negative scale on the group, a flip operation on the group, will flip the contained object.
Why in that case you would remove the obtained flip?
In that case it would be a user driven flip

**Comment by melvynhills** [comment]
@asturur No, I'm not sure, especially for edge cases with both scaling & flip. This patch could help some people with basic use cases.
I will update this thread with future improvements.

**Comment by asturur** [comment]
a way to test it is to check if object.calcTransformMatrix() after your code is the same that you obtain before running your code.

**Comment by asturur** [comment]
To be clear, if someone wants to find a solution is encouraged to do so.
I ll flag this as possible feature and i will leave it open.

**Comment by asturur** [comment]
I do think the problem is complicated both code wise and geometrically, i do not think is a quick fix.
